### PR TITLE
Add Reinterpret operator as a flavor of Reshape

### PR DESCRIPTION
**Pull request by mzient**
Add Reinterpret operator as a flavor of Reshape
* `dtype` argument specifies target type
* Output shape is adjusted by default in the innermost dimension
* Wildcard shape calculation includes the type size ratio
* Add tests

Changes required to make it work:
* Add `Resize(new_shape, new_type)` to TensorList, TensorVector and Tensor and Buffer

Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature needed to reinterpret TFRecord data

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
    Add Reinterpret operator as a flavor of Reshape
      * `dtype` argument specifies target type
      * Output shape is adjusted by default in the innermost dimension
      * Wildcard shape calculation includes the type size ratio
      * Add tests
    Changes required to make it work:
      * Add `Resize(new_shape, new_type)` to TensorList, TensorVector and Tensor and Buffer
       - Affected modules and functionalities:
 - Key points relevant for the review:
     * ?
 - Validation and testing:
     * python test
 - Documentation (including examples):
     * docstring

**JIRA TASK**: DALI-1301

### CODE REVIEW

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
> Why we need another operator and can't just extend current reshape?

I think that making `Reshape` also change the type is a bit of a stretch - and it's not something I'd even look for as a user. `Reinterpret` pretty much says it all.
Numpy has separate `view` and `reshape` functions.

File: dali/operators/generic/reshape.cc [hunk 1]
```diff
@@ -219,9 +251,14 @@ void Reshape<Backend>::CalculateOutputShape(const Workspace &ws) {
       break;
     case ShapeSource::None:
       output_shape_ = input_shape_;
-      return;
+      break;
   }
 
+  int64_t input_element_size = in.type().size();
```
Comment (jantonguirao) [inline dali/operators/generic/reshape.cc]: a suggestion: You could check output_type_ earlier and later just assume that output_type_ is a valid type
```
if (output_type_ != DALI_NO_TYPE)
    output_type_ = in.type();
```

File: dali/test/python/test_operator_reshape.py [hunk 2]
```diff
@@ -207,13 +207,101 @@ def test_reshape_arg_input():
         for use_wildcard in [False, True]:
            yield check_reshape_with_arg_input, device, batch_size, relative, use_wildcard
 
+class ReinterpretPipeline1(Pipeline):
```
Comment (jantonguirao) [inline dali/test/python/test_operator_reshape.py]: a quick note of what is this (`ReinterpretPipeline1` and `ReinterpretPipeline2` doesn't tell me much) would help to read this code

**Comment by mzient** [comment]
!build

File: dali/operators/generic/reshape.cc (hunk 1 above)
Comment (mzient) [reply dali/operators/generic/reshape.cc]: Done.

**Pull request merged**